core/utils/trie_test.cc

```diff
@@ -1,91 +1,190 @@
+  EXPECT_FALSE(trie.Lookup("Hello").first);
+  EXPECT_FALSE(trie.Lookup("H").first);
+  EXPECT_FALSE(trie.Lookup("").first);
+}
+
+// Whether an Trie with the "" prefix behaves correctly
+TEST(TrieTest, EmptyPrefix) {
+  Trie<int32_t> trie;
+  trie.Insert("", 2, true);
 
-  EXPECT_FALSE(trie.Lookup("234"));
-  EXPECT_FALSE(trie.Lookup("ello"));
-  EXPECT_FALSE(trie.Lookup("Hello"));
-  EXPECT_FALSE(trie.Lookup("H"));
+  EXPECT_TRUE(trie.Match(""));
+  EXPECT_TRUE(trie.Match("234"));
+  EXPECT_TRUE(trie.Match("ello"));
+  EXPECT_TRUE(trie.Match("Hello"));
+  EXPECT_TRUE(trie.Match("H"));
+  EXPECT_TRUE(trie.MatchPrefix("Hello"));
+  EXPECT_TRUE(trie.MatchPrefix(""));
 
-  EXPECT_TRUE(trie.Lookup(""));
+  EXPECT_TRUE(trie.Lookup("Hello").first);
+  EXPECT_EQ(trie.Lookup("Hello").second, 2);
 
-  EXPECT_FALSE(trie.LookupKey("234"));
-  EXPECT_FALSE(trie.LookupKey("ello"));
-  EXPECT_FALSE(trie.LookupKey("Hello"));
-  EXPECT_FALSE(trie.LookupKey("H"));
-  EXPECT_FALSE(trie.LookupKey(""));
+  EXPECT_TRUE(trie.Lookup("H").first);
+  EXPECT_EQ(trie.Lookup("H").second, 2);
 }
 
 // Test the copy constructor
 TEST(TrieTest, Copy) {
-  Trie trie0;
-
-  Trie trie1 = trie0;
-  EXPECT_FALSE(trie1.Lookup("Hello"));
-  EXPECT_FALSE(trie1.Lookup("H"));
-  EXPECT_TRUE(trie1.Lookup(""));
-  EXPECT_FALSE(trie1.LookupKey("234"));
-  EXPECT_FALSE(trie1.LookupKey("ello"));
-  EXPECT_FALSE(trie1.LookupKey(""));
-
-  trie0.Insert("Hello world!");
-  trie0.Insert("123456");
-  Trie trie2 = trie0;
-
-  EXPECT_FALSE(trie2.Lookup("234"));
-  EXPECT_FALSE(trie2.Lookup("ello"));
-  EXPECT_TRUE(trie2.Lookup("Hello"));
-  EXPECT_TRUE(trie2.Lookup("H"));
-  EXPECT_TRUE(trie2.Lookup(""));
-  EXPECT_TRUE(trie2.Lookup("1"));
-  EXPECT_TRUE(trie2.Lookup("123456"));
-  EXPECT_FALSE(trie2.LookupKey("Hello"));
-  EXPECT_FALSE(trie2.LookupKey("1"));
-  EXPECT_TRUE(trie2.LookupKey("Hello world!"));
-  EXPECT_TRUE(trie2.LookupKey("123456"));
-  EXPECT_TRUE(trie2.LookupKey("1234567"));
+  Trie<int32_t> trie0;
+
+  Trie<int32_t> trie1 = trie0;
+  EXPECT_FALSE(trie1.Match("Hello"));
+  EXPECT_FALSE(trie1.Match("H"));
+  EXPECT_FALSE(trie1.Match(""));
+  EXPECT_FALSE(trie1.MatchPrefix("234"));
+  EXPECT_FALSE(trie1.MatchPrefix("ello"));
+  EXPECT_TRUE(trie1.MatchPrefix(""));
+
+  trie0.Insert("Hello world!", 1);
+  trie0.Insert("Hello", 2, true);
+  trie0.Insert("123456", 3);
+  Trie<int32_t> trie2 = trie0;
+
+  EXPECT_FALSE(trie2.Match("234"));
+  EXPECT_FALSE(trie2.Match("ello"));
+
+  EXPECT_TRUE(trie2.Match("Hello"));
+  EXPECT_TRUE(trie2.Match("Hello y'all"));
+  EXPECT_TRUE(trie2.MatchPrefix("H"));
+  EXPECT_TRUE(trie2.MatchPrefix(""));
+  EXPECT_TRUE(trie2.MatchPrefix("1"));
+
+  EXPECT_TRUE(trie2.Lookup("Hello").first);
+  EXPECT_TRUE(trie2.Lookup("Hello y'all").first);
+  EXPECT_TRUE(trie2.Lookup("Hello World").first);
+
+  EXPECT_EQ(trie2.Lookup("Hello").second, 2);
+  EXPECT_EQ(trie2.Lookup("Hello y'all").second, 2);
+  EXPECT_EQ(trie2.Lookup("Hello World").second, 2);
```

Comment:
The interesting corner cases that arise with respect to the management of the Trie are inserts and deletes which force the creation and destruction of Nodes and the associated processing of the children array. A test that randomly creates thousands of keys of say between 1 and 8 characters and then randomly removes and inserts keys while ensuring that lookups work as expected might more thoroughly prove out these algorithms.